### DISCUSSION TITLE
UDP/ DTLS decryption

### DISCUSSION THREAD

**Discussion by SoerenSteinberg** [mitmproxy/mitmproxy#6268] [Q&A] [answered]
Hi, I have a question about UDP/Wireguard mode. If I use the mode and route UDP/DTLS traffic through the proxy, the UDP packets are still displayed encrypted. Is an insight into plain text currently not supported?
Thank you in advance!

**Comment by mhils** [comment, marked as answer]
Generally speaking that should work. I've lately refactored the detection logic here: https://github.com/mitmproxy/mitmproxy/blob/b14f9611f248df8ecfec9b5d1d556ce6e22ec64c/mitmproxy/addons/next_layer.py#L149-L158

Would you mind installing from main (or from a dev snapshot produced by [CI](https://github.com/mitmproxy/mitmproxy/actions/workflows/main.yml?query=branch%3Amain)) and trying again? If it's not picked up, a log with `-vvv --set proxy_debug` would be helpful.

**Reply by SoerenSteinberg** [reply]
Many thanks for the quick response! I have now installed mitmproxy in version 10.0.0.dev of Git and unfortunately the problem is still there. I'm trying to route a CoAP message (DTLS encrypted) through the proxy. Could it be the protocol? Here is the anonymized log file. Thank you in advance for the help!

[debug.log](https://github.com/mitmproxy/mitmproxy/files/12137009/debug.log)

**Reply by mhils** [reply]
Thanks! Looking at the log file, it seems like your messages use DTLS 1.0. (first message is 0x16 + 0xfeff), which makes `starts_like_dtls_record` return false:
https://github.com/mitmproxy/mitmproxy/blob/cd4c13a21c8168adb3afb6335a3725ffd1d2cc9a/mitmproxy/net/tls.py#L253-L264.

I haven't looked into DTLS 1.0 at all - maybe just replace the last `0xFE` with `0xFF` in that snippet and see what happens? If things work, I'd appreciate a PR that bumps it for everyone! 😃

**Reply by SoerenSteinberg** [reply]
Wow, thanks for the quick reply! However, I'm a little surprised about the alleged use of DTLS 1.0. If I record the traffic with Wireshark without the MITMProxy, DTLS 1.2 is definitely used. I tested your change once. In any case, the handshake is started by the change. However, I get the following error:

> _21:37:54.792][10.0.0.1:60845] Unable to establish DTLS connection with server (Certificate verify failed: IP address mismatch). Trying to establish DTLS with client anyway. If you plan to redirect requests away from this server, consider setting `connection_strategy` to `lazy` to suppress early connections.
> [21:37:54.795][10.0.0.1:60845] Client TLS handshake failed. The client may not trust the proxy's certificate for IP:5684 (OpenSSL Error([('SSL routines', '', 'no shared cipher')]))_

The MITMProxy certificate is correctly installed and valid on the client. Do you happen to have an idea as to what could be causing this?

Thank you in advance!

**Reply by mhils** [reply]
> 21:37:54.792][10.0.0.1:60845] Unable to establish DTLS connection with server (Certificate verify failed: IP address mismatch). Trying to establish DTLS with client anyway. If you plan to redirect requests away from this server, consider setting connection_strategy to lazy to suppress early connections.

mitmproxy is not trusting the upstream certificate. You can temporarily enable the `ssl_insecure` option to work around that.

> [21:37:54.795][10.0.0.1:60845] Client TLS handshake failed. The client may not trust the proxy's certificate for IP:5684 (OpenSSL Error([('SSL routines', '', 'no shared cipher')]))

mitmproxy and client can't agree on a cipher. Setting [`ciphers_client`](https://docs.mitmproxy.org/stable/concepts-options/#ciphers_client) to something like `ALL` may fix it. If that does not work, more dedicated low-level debugging is required. :)

**Reply by SoerenSteinberg** [reply]
thank you again for the quick reply. Unfortunately, setting the cipher_suite to All didn't change anything. I've added a Dubug log to the system. Could you here check it again when you have time? maybe do you have any idea how to solve this problem? Thank you in advance!
[debug.log](https://github.com/mitmproxy/mitmproxy/files/12169118/debug.log)

**Reply by mhils** [reply]
From your log:

> Client TLS handshake failed. The client may not trust the proxy's certificate for IP:5684

This points at one of two things:

1. Your client does not trust mitmproxy's CA certificate
2. There are some other TLS handshake incompatibilities / intricacies that require a deep dive.

**Comment by SoerenSteinberg** [comment]
That is very strange. From my point of view, the certificate is correctly installed and valid. Can it be that mitmproxy does not support the following ciphersuites?

- TLS_PSK_WITH_AES_128_GCM_SHA256
- TLS_PSK_WITH_AES_256_GCM_SHA384
- TLS_PSK_WITH_AES_128_CCM_8
- TLS_PSK_WITH_AES_256_CCM_8
- TLS_PSK_WITH_AES_128_CCM
- TLS_PSK_WITH_AES_256_CCM
- TLS_PSK_WITH_AES_128_CBC_SHA256

**Reply by mhils** [reply]
Oh - those are all PSK ciphersuites, which means you need to communicate the pre-shared key to OpenSSL in advance. I don't think we have any APIs to do this in mitmproxy (though contributions are welcome).
